Added some kestrel event counters

davidfowl opened the pull request:
## Connection counters
- Connection queue length - This is the amount of connections accepted and queued in the thread pool.
- Connection count - The number of connections
- Total connections - The total number of connections.
- Connection Rate - Connections per second

## TLS counters
- Total TLS handshakes - The total number of TLS handshakes
- Current TLS handshakes - The number of TLS handshakes being performed currently
- Failed TLS handshakes - The number of failed TLS handshakes
- TLS handshake Rate - TLS handshake per second

## Request counters
- Request Queue Length - The number of HTTP/2 and HTTP/3 connections queued in the thread pool.
- Current Upgraded Requests - The number of HTTP/1 requests that are currently upgraded (basiacally the number of concurrent websocket connections)

- Also added event source events to Kestrel (which I'm beginning to get very conflicted about 😄 )

Contributes to #4769

Notes: The Request Queue Length only works for HTTP/2 and HTTP/3. The Connection Queue doubles as the request queue for HTTP/1.

cc @shirhatti

davidfowl committed:
Added some kestrel event counters
- Connection queue length - This is the amount of connections accepted and queued in the thread pool.
- Connection count - The number of connections
- Total connections - The total number of connections ever connected.
- Connection Rate - Connections per second

Contributes to #4769

davidfowl committed:
Copy pasta

davidfowl committed:
Added TLS counters
- Current TLS handshakes
- Total TLS handshakes
- Failed TLS handshakes
- TLS handshake per second

davidfowl committed:
Added HTTP/2 queue length counter

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs:
```diff
@@ -261,6 +299,11 @@ protected override void OnEventCommand(EventCommandEventArgs command)
                     DisplayName = "Connection Queue Length"
                 };
+
+                _http2RequestQueueLengthCounter ??= new PollingCounter("http2-request-queue-length", this, () => _http2RequestQueueLength)
```
Should HTTP/2 in the counter name? If HTTP/3 has the same architecture then a neutral name would allow the counter to be reused.

davidfowl replied:
I was looking to see if I could do the same for all HTTP versions but it might make sense to have a counter for each since we don’t have dimensions

davidfowl committed:
Improve the event information
- Add TLS version to handshake events
- Add HTTP version to request queue events
- Renamed HTTP/2 request queue length to http request queue

davidfowl committed:
Add the host name to the TLS handshake stop

davidfowl committed:
Clean up
- Event sources don't like null values
- Make the level information for most events

davidfowl committed:
Added counter for upgraded connections

davidfowl committed:
More tweaks
- Add more data to request events

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2StreamOfT.cs:
```diff
@@ -19,6 +20,7 @@ public Http2Stream(IHttpApplication<TContext> application, Http2StreamContext co
 
         public override void Execute()
         {
+            KestrelEventSource.Log.RequestDequeued(this, AspNetCore.Http.HttpProtocol.Http2);
```
I'm thinking of moving this after request parsing. It complicates things a bit though in case that logic fails.

halter73 replied:
We could just invoke Http2Stream.TryValidatePseudoHeaders() before dispatching to the thread pool.

davidfowl replied:
I'll do that in another PR.

halter73 commented on src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:
```diff
@@ -169,6 +169,7 @@ public async Task OnConnectionAsync(ConnectionContext context)
                     {
                         selector = (sender, name) =>
                         {
+                            feature.HostName = name;
```
Should write log the hostname when a single cert is used and not an SNI callback meaning `_serverCertificateSelector` is null? We could always pass in a ServerCertificateSelectionCallback and just have it return the `_serverCertificate` when no custom callback is configured.

Also, don't you love it when you thought you left a comment days ago, but you accidentally started a review you didn't finsih?

davidfowl replied:
> Should write log the hostname when a single cert is used and not an SNI callback meaning _serverCertificateSelector is null? We could always pass in a ServerCertificateSelectionCallback and just have it return the _serverCertificate when no custom callback is configured.

Where do I get that information? Is it on SSlStream?

halter73 replied:
> Where do I get that information? Is it on SSlStream?

I don't know. I don't think so. That's why I suggested always using a ServerCertificateSelectionCallback, so the hostname could be captured inside the callback.

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs:
```diff
@@ -1002,6 +1002,7 @@ private void StartStream()
                 throw;
             }
 
+            KestrelEventSource.Log.RequestQueued(_currentHeadersStream, AspNetCore.Http.HttpProtocol.Http2);
```
While it's just another work item, I think this could be useful to quantify just how backlogged the thread pool is. It's one thing to say there's a thousand pending work items, but another to say there are thousand pending requests.

davidfowl replied:
It's literally called `QueueUserWorkitem` 😄 , it's absolutely a queue. We're just counting how may items are ours vs global in the pool like @halter73 says.

Tratcher replied:
The same request will end up back in the same threadpool queue as soon as some middleware goes async. What's special about this specific instance? I don't see this producing any actionable data.

Tratcher replied:
Have you produced any graphs with this yet? I'm really curious what they'd show.

davidfowl replied:
> The same request will end up back in the same threadpool queue as soon as some middleware goes async. 

That's not really true. Actually by default continuations will go into a different queue (well thread local stack). 

> What's special about this specific instance? I don't see this producing any actionable data.

It's useful to see this number be non-zero. While it doesn't capture the entire end to end async flow of the request it can give a sense of things getting held up outside global threadpool queue.

shirhatti replied:
> It's useful to see this number be non-zero.

Under sustained load, wouldn't this always be non-zero?

Under what conditions do you expect this number to diverge from it's usual value?

davidfowl replied:
> Under sustained load, wouldn't this always be non-zero?

It'll be non-zero if there's blocking or a *huge* amount of load.

> Under what conditions do you expect this number to diverge from it's usual value?

Starvation/blocking and it'll show up under a thing called connections/requests which is why I like it more than just having a single "threadpool queue" counter (though that is useful as well).

shirhatti replied:
For a new request on an existing connection, I'm assuming this is the first threadpool dispatch. However. in a threadpool starved scenario, this won't be the first threadpool dispatch for a request on a new connection right?

Am I correct in understanding that only the first scenario (i.e., requests on existing connections) will cause this counter to diverge from zero?

davidfowl replied:
For HTTP/1 the request queue is the connection queue for HTTP/2 and 3 we queue per request

halter73 replied:
This got me thinking maybe we should add a PollingCounter for concurrent idle/active connections (e.g. connections (not) in the keep-alive state). It wouldn't be equivalent to `"current-connections" - "current-requests"` because of HTTP/2 and 3. It might even be different for HTTP/1.x if we consider a connection "active" as soon as we get any part of the request line and/or headers.

shirhatti commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs:
```diff
@@ -79,37 +141,235 @@ public void ConnectionRejected(string connectionId)
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [Event(6, Level = EventLevel.Informational)]
+        private void ConnectionQueued(string connectionId,
```
Should we change this Event Name? As you mentioned offline suffixing with Start/Stop gives semantic meaning to Events?

davidfowl replied:
ConnectionQueuedStart?

halter73 replied:
While "Dequeued" is a little more explicit that "Stop", hopefully it will be obvious enough that it's the time in the queue that's being measured, not just the short amount of time it takes to enqueue the connection/request.

davidfowl replied:
Will do.

davidfowl committed:
PR feedback

shirhatti commented:
@sebastienros We just added new EventCounters. I've filed an issue for having these enabled in our benchmarks https://github.com/aspnet/Benchmarks/issues/1513

davidsh commented:
Added new issue https://github.com/dotnet/diagnostics/issues/1188 to make sure 'dotnet-counters' tool is updated for these new counters.
